### PR TITLE
ERC20 Token Subgraph and Analytics APP

### PR DESCRIPTION
**Pull request by corerouter**
Check in ERC20 token subgraph and Analytics APP for review. Thanks!

### CODE REVIEW

File: subgraphs/erc20/README.md [hunk 1]
```diff
@@ -1 +1,27 @@
-# ERC20-Subgraph
\ No newline at end of file
+# ERC20 Token Subgraph
+The subgraph read erc20 token list from IPFS, monitor the tokens' event and aggregate the results accordingly.
+
+## Calculation Methodology v1.0.0
+
+### Token's Hold Count
```
Comment (this-username-is-taken) [inline subgraphs/erc20/README.md:6]: `Holder Count`

File: subgraphs/erc20/README.md [hunk 2]
```diff
@@ -1 +1,27 @@
...
+### Token's Transfer Count
+
+Count of Transfer Event which have interacted with the protocol via transfer transaction
+
+`Transfers`
+
+### Token's Cumulative Unique Users
+Count of Unique Addresses which have interacted with the protocol via any transaction
+
+`Transfers`
```
Comment (this-username-is-taken) [inline subgraphs/erc20/README.md:19]: what's this `transfer` for?

**Comment by this-username-is-taken** [comment]
Can you use `prettier` to format the code?

File: subgraphs/erc20/src/mappings/registry.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+export function createToken(value: JSONValue, userData: Value): void {
+
+  
+  let rawData = value.toArray() 
+    
+  let address = rawData[0].isNull() ? "" : rawData[0].toString()
+  let name = rawData[1].isNull() ? "" : rawData[1].toString()
+  let symbol = rawData[2].isNull() ? "" : rawData[2].toString()
```
Comment (this-username-is-taken) [inline subgraphs/erc20/src/mappings/registry.ts:111]: why not get these from the smart contracts?

File: subgraphs/erc20/schema.graphql [hunk 4]
```diff
@@ -1,269 +1,285 @@
...
+  name: String!
+
+  " Symbol of the token, mirrored from the smart contract "
+  symbol: String!
+
+  " The number of decimal places this token uses, default to 18 "
+  decimals: Int!
+
+  " Token flags"
+  flags: [String!]!
```
Comment (this-username-is-taken) [inline subgraphs/erc20/schema.graphql:22]: why not make this an Enum?

File: subgraphs/erc20/README.md (hunk 1 above)
Comment (corerouter) [reply subgraphs/erc20/README.md]: changed in the latest commit.

File: subgraphs/erc20/README.md (hunk 2 above)
Comment (corerouter) [reply subgraphs/erc20/README.md:19]: change to Holder Count in the latest commit.

File: subgraphs/erc20/src/common/constants.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+export const SECONDS_PER_HOUR = 60 * 60;
+
+export const BIGINT_ZERO = BigInt.fromI32(0);
+export const BIGINT_ONE = BigInt.fromI32(1);
+export const BIGINT_TWO = BigInt.fromI32(2);
+export const BIGDECIMAL_ZERO = new BigDecimal(BIGINT_ZERO);
+export const BIGDECIMAL_ONE = new BigDecimal(BIGINT_ONE);
+
+// Flags
+export const DETAILED_TOKEN = 1 << 0;
```
Comment (this-username-is-taken) [inline subgraphs/erc20/src/common/constants.ts:19]: Where is this used?

File: subgraphs/erc20/src/mappings/token.ts [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+
+    let isBurn = event.params.to.toHex() == GENESIS_ADDRESS;
+    let isMint = event.params.from.toHex() == GENESIS_ADDRESS;
+    let isTransfer = !isBurn && !isMint;
+
+    if (isBurn) {
+      handleBurnEvent(token, amount, event.params.from, event);
+    } else if (isMint) {
+      log.error("isMint", []);
+      handleMintEvent(token, amount, event.params.to, event);
```
Comment (harsh9200) [inline subgraphs/erc20/src/mappings/token.ts:75]: If a mint/burn event is emitted with a transfer event, won't the `handleBurnEvent`/`handleMintEvent` functions run twice?  shouldn't we add a condition for this?

File: subgraphs/erc20/src/common/constants.ts (hunk 5 above)
Comment (corerouter) [reply subgraphs/erc20/src/common/constants.ts]: It was designed to be a flag to indicate whether an erc20 token supports functions such as name(), symbol(). After consideration, I decided to disable this flag as when there are several hundred erc20 tokens to index there is no way to     know it in advance and we just try to call these methods later.

File: subgraphs/erc20/src/mappings/token.ts (hunk 6 above)
Comment (corerouter) [reply subgraphs/erc20/src/mappings/token.ts]: Yes. In some library a special transfer function  emit both transfer and mint/burn event. I checked whether on chain totalSupply has been updated to decides the event related with the transfer call has been processed before in today's commit https://github.com/messari/subgraphs/pull/546/commits/c877e3aa03afd264dbb14be0131529662edbc13c Thanks!

**Comment by this-username-is-taken** [comment]
You should put the erc721 subgraph in a separate PR

File: subgraphs/erc721/schema.graphql [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+  name: String!
+
+  " Symbol of the smart contract "
+  symbol: String!
+
+  " Flag for whether EIP721 metadata is supported "
+  supportsEIP721Metadata: Boolean!
+
+  " Total number of unique tokens in the smart contract "
+  tokenCount: BigInt!
```
Comment (0xbe1) [inline subgraphs/erc721/schema.graphql:47]: do you mean minted token count?

**Comment by corerouter** [comment]
> PR

Done. Create a new PR at https://github.com/messari/subgraphs/pull/689/commits

File: subgraphs/erc721/schema.graphql [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+#############################################
+####### ERC721 Smart Contract Metadata ######
+#############################################
+
+type Contract @entity {
+  " ERC721 Smart contract address "
+  id: ID!
+
+  " Name of the smart contract "
+  name: String!
```
[the remark this answers is not in the document]
Comment (corerouter) [reply subgraphs/erc721/schema.graphql]: Changed. Please see an updated version at https://github.com/messari/subgraphs/pull/689/commits

File: subgraphs/erc721/schema.graphql [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+  " Symbol of the smart contract "
+  symbol: String!
```
[the remark this answers is not in the document]
Comment (corerouter) [reply subgraphs/erc721/schema.graphql]: Changed. Please see an updated version at https://github.com/messari/subgraphs/pull/689/commits

File: subgraphs/erc721/schema.graphql [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+  " Flag for whether EIP721 metadata is supported "
```
[the remark this answers is not in the document]
Comment (corerouter) [reply subgraphs/erc721/schema.graphql]: Changed. Please see an updated version at https://github.com/messari/subgraphs/pull/689/commits

File: subgraphs/erc721/schema.graphql [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+#############################
+####### Token Metadata ######
+#############################
+
+type Token @entity {
+  " address of the token, { Contract Address }-{ Token ID }  "
+  id: ID!
+
+  " Smart contract address for the token "
```
[the remark this answers is not in the document]
Comment (corerouter) [reply subgraphs/erc721/schema.graphql]: Changed. Please see an updated version at https://github.com/messari/subgraphs/pull/689/commits

File: subgraphs/erc721/schema.graphql [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+  contract: Contract!
+
+  " ID of the token from the smart contract "
+  tokenID: BigInt!
+
+  " URI of the token "
+  tokenURI: String!
```
[the remark this answers is not in the document]
Comment (corerouter) [reply subgraphs/erc721/schema.graphql]: Changed. Please see an updated version at https://github.com/messari/subgraphs/pull/689/commits

File: subgraphs/erc721/schema.graphql [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+
+  " The owner's address for the token "
```
[the remark this answers is not in the document]
Comment (corerouter) [reply subgraphs/erc721/schema.graphql]: Changed. Please see an updated version at https://github.com/messari/subgraphs/pull/689/commits

File: subgraphs/erc721/schema.graphql (hunk 7 above)
Comment (corerouter) [reply subgraphs/erc721/schema.graphql]: Yes. Update the comment part accordingly. Please see an updated version at https://github.com/messari/subgraphs/pull/689/commits

**Comment by this-username-is-taken** [comment]
Awesome job on this!

**Pull request merged**